### PR TITLE
Fix column naming edge cases - invalid and duplicated columns, case-insensitive name aliasing for case-insensitive backends

### LINKED ISSUE #7412: Unexpected behaviour in SQLite queries due to case insensitivity of the database
### Intro

The SQLite database is not case-sensitive for column names, but our implementation is case-sensitive.

That means that Enso allows to create queries with two columns that differ only in case, e.g. `x` and `X`. For SQLite these are indistinguishable and if they show up in queries (as aliases), they can make the behaviour completely incorrect. The worst thing is - in case of ambiguous references - unlike Postgres which will fail the query, SQLite just resolves the first column that matches - leading to data corruption.

### Repro

To illustrate:
```python
from Standard.Base import all
from Standard.Table import all
from Standard.Database import all

main =
    c = Database.connect (SQLite In_Memory)
    t0 = Table.new [["X", [1]]]
    t1 = t0.select_into_database_table c "TABLE1"

    t2 = t1.set "[X] + 100" "x"
    IO.println "t2:"
    IO.println t2.to_sql.prepare.first
    t2.print

    t3 = t2.join t2 on="X"
    IO.println "t3:"
    t3.print

    t4 = t2.union t2
    IO.println "t4:"
    IO.println t4.to_sql.prepare.first
    t4.print

    t5 = t2.reorder_columns "x"
    IO.println "t5:"
    t5.print

    t6 = t5.join t5 on="X"
    IO.println "t6:"
    t6.print

    t7 = t5.union t5
    IO.println "t7:"
    t7.print
```

### Actual behaviour
Currently yields
```
t2:
SELECT "TABLE1"."X" AS "X", ("TABLE1"."X" + ?) AS "x" FROM "TABLE1" AS "TABLE1"
 X | x
---+-----
 1 | 101

t3:
 X | x | Right X | Right x
---+---+---------+---------
 1 | 1 | 1       | 1

t4:
SELECT CAST("TABLE1_TABLE1"."X" AS INTEGER) AS "X", CAST("TABLE1_TABLE1"."x" AS INTEGER) AS "x" FROM (SELECT "TABLE1"."X" AS "X", ("TABLE1"."X" + ?) AS "x" FROM "TABLE1" AS "TABLE1" UNION ALL SELECT "TABLE1"."X" AS "X", ("TABLE1"."X" + ?) AS "x" FROM "TABLE1" AS "TABLE1") AS "TABLE1_TABLE1"
 X | x
---+---
 1 | 1
 1 | 1

t5:
 x   | X
-----+---
 101 | 1

t6:
 x   | X   | Right x | Right X
-----+-----+---------+---------
 101 | 101 | 101     | 101

t7:
 x   | X
-----+-----
 101 | 101
 101 | 101
```

We create a new row `x` in `t2` by computing `[X] + 100` (in the SQL this is `"X" + ?` because the `?` is an interpolated parameter, filled in when executing the query). So far so good - the table is displayed correctly - in the only row column `X` has value `1` and the computed `x` has value `101`.

However, once we do any operations like `join` or `union` we can see that the value `101` **disappears** and is replaced with `1` instead!!

Looking at the SQL code generated for `t4` we can see why that may be - the new query is based on a subquery for the union - and in the subquery we have columns named `X` and `x`. Apparently, due to case insensitivity, SQLite when resolving both `"TABLE1_TABLE1"."X"` and `"TABLE1_TABLE1"."x"` is happy to go with the first matching column, yielding the same value for both columns - even though they should be different.

In `t5` we reorder the elements and do the same queries - now we can see that both values got replaced with `101`! That's probably because now that is the first matching column.

This is very concerning - of course such aliasing of names just differing in case may not be most common - but whenever it occurs, the computed data may be completely off and behave in inpredictable ways - as demonstrated in this example - the results are different depending on the order of columns.

### Expected behaviour

If the computation were to be computed correctly, e.g. if I replace the SQLite backend with Postgres, we will get:

```
t2:
SELECT "TABLE1"."X" AS "X", ("TABLE1"."X" + ?) AS "x" FROM "TABLE1" AS "TABLE1"
 X | x
---+-----
 1 | 101

t3:
 X | x   | Right X | Right x
---+-----+---------+---------
 1 | 101 | 1       | 101

t4:
SELECT "TABLE1_TABLE1"."X" AS "X", "TABLE1_TABLE1"."x" AS "x" FROM (SELECT "TABLE1"."X" AS "X", ("TABLE1"."X" + ?) AS "x" FROM "TABLE1" AS "TABLE1" UNION ALL SELECT "TABLE1"."X" AS "X", ("TABLE1"."X" + ?) AS "x" FROM "TABLE1" AS "TABLE1") AS "TABLE1_TABLE1"
 X | x
---+-----
 1 | 101
 1 | 101

t5:
 x   | X
-----+---
 101 | 1

t6:
 x   | X | Right x | Right X
-----+---+---------+---------
 101 | 1 | 101     | 1

t7:
 x   | X
-----+---
 101 | 1
 101 | 1
```

Here we can see that in all instances `X` is staying `1` and `x` is staying `101` - the columns are correctly distinguished.

This may not be directly achievable with SQLite - since it's just case insensitive we cannot get that. We could internally rename the columns to distinguish them, but that's rather not aligned with our philosophy of 'try to keep column names visible to the user aligned with what happens in SQL' (e.g. how we are truncating column names in #5951). 

Instead, it seems that the correct behaviour would be to **not allow renaming columns to alias each other case insensitively**. I.e. if I have a column `X` in a table and try to add a column `x` to it, I should get some kind of error telling me that the names are aliased.

Whether this should be done for all backends or just the ones that handle names case-insensitively - I'm not sure - both approaches have some + and -. It would be nice to have consistent behaviour across all backends, so it may be worth to make names case-insensitive everywhere, but that may be a bit too drastic change. Alternatively, we can detect this and only error on backends where it is an issue (e.g. SQLite).

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Fixes #7412 
- Also adds tests and fixes some more edge cases:
	- Ensures correct handling of existing Database tables whose column names may be invalid from Enso perspective, or clashing from Enso perspective (e.g. for most DBs `ś` and `s\u0301` are different names, but for Enso they are basically the same so this would cause issues - thus Enso now renames such columns when accessed (still using the correct column reference in the generated SQL under the hood).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Database/0.0.0-dev/src/Connection/Connection.enso [hunk 1]
```diff
@@ -221,12 +221,23 @@ type Connection
...
+                   data corruption. Our renaming mechanism is used to fix issues
+                   with existing tables with problematic names, but that
+                   mechanism will only provide a false-sense of security in case
+                   of a query, whereas a query like `SELECT 1 AS "A", 2 AS "A"`
+                   will actually result in both columns `A` and `A 1` containing
+                   the value 1; and value 2 being lost. That is why such queries
+                   must fail.
+                r = Database_Table_Module.make_table self name columns ctx on_problems=Problem_Behavior.Report_Error
+                r.catch Any error->
+                    Error.throw (Illegal_Argument.Error "The provided custom SQL query is invalid and may suffer data corruption when being processed, especially if it contains aliased column names, it may not be interpreted correctly. Please ensure the names are unique. The original error was: "+error.to_display_text cause=error)
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Connection/Connection.enso:240]: This is a good safeguard, because raw SQL queries could have problems we can't forsee.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Entity_Naming_Properties.enso [hunk 2]
```diff
@@ -33,11 +33,14 @@ new jdbc_connection =
         if column_limit == 0 then
             Panic.throw (Illegal_State.Error "Unexpected: The database server does not report the maximum column name length.")
 
-        table_properties = Encoding_Limited_Naming_Properties.Instance encoding table_limit
-        column_properties = Encoding_Limited_Naming_Properties.Instance encoding column_limit
+        ## Postgres column/table names may be lowercased if unquoted, but when
+           quoted we have full case sensitivity and can distinguish columns
+           `A` from `a` as different.
+           Our generator is supposed to always quote identifiers.
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Entity_Naming_Properties.enso:39]: Could this cause a problem if the user includes raw SQL in a larger query, and expects case insensitivity, in a join for example?

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Entity_Naming_Properties.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Entity_Naming_Properties.enso:39]: I don't think I understand the question.

What do you mean by the user expecting case insensitivity? Could you provide some example query?
